dashboard/app/helpers/multiple_databases_transition_helper.rb

```diff
@@ -0,0 +1,72 @@
+require "active_support/concern"
+
+# A helper module to factilitate the transition from SeamlessDatabasePool to
+# ActiveRecord's new Multiple Databases implementation.
+module MultipleDatabasesTransitionHelper
+  # Provide a new `use_reader_connection_for_route` method on controllers to
+  # replace our existing use of `use_database_pool route_name: :persistent`
+  # which can be used to declare that a particular route should explicitly
+  # always use the read replica.
+  module ControllerFilter
+    extend ActiveSupport::Concern
+    class_methods do
+      def use_reader_connection_for_route(route)
+        if MultipleDatabasesTransitionHelper.transitioned?
+          around_action :connect_to_reader, only: route
+        else
+          use_database_pool route => :persistent
+        end
+      end
+    end
+
+    def connect_to_reader
+      ActiveRecord::Base.connected_to(role: :reading) do
+        yield
+      end
+    end
+  end
+
+  # Provide a new `use_writer_connection` method to replace our existing use of
+  # `use_master_connection` with the entirely-equivalent ActiveRecord
+  # implementation.
+  def self.use_writer_connection
+    if MultipleDatabasesTransitionHelper.transitioned?
+      ActiveRecord::Base.connected_to(role: :writing) do
+        yield
+      end
+    else
+      SeamlessDatabasePool.use_master_connection do
+        yield
+      end
+    end
+  end
+
+  # Provide a new `use_persistent_read` method to replace our existing use of
+  # the same. ActiveRecord itself has some new logic to provide similar
+  # functionality, as does ProxySQL, so it could well be that we don't actually
+  # need this at all post-transition. More investigation needed for this TODO
+  # to be resolved.
+  def self.use_persistent_read_connection
+    if MultipleDatabasesTransitionHelper.transitioned?
+      # TODO
+    else
+      SeamlessDatabasePool.use_persistent_read_connection do
+        yield
+      end
+    end
+  end
+
+  # Whether we are still on the old version (false, pre-transition) or the new
+  # version (true, post-transition). Hardcoded to the known versions for
+  # resiliency and narrow focus.
+  def self.transitioned?
+    case Rails.version
+    when "5.2.4.4"
+      false
+    when "6.0.4.1"
+      true
```

Comment:
love this - I was expecting to just see `transitioned = false`

Reply:
Is there a possibility we will patch Rails 5.2 before upgrading to Rails 6 or that we would adopt a slightly different version of Rails 6? Wondering whether we should make the check less specific and distinguish between `5.2.*` vs `6.0.*`

Reply:
theoretically, yes, but I'm not confident in the ability of this code to support other versions of Rails; also, the idea is that this code should not need to exist for long enough for that to be relevant. If we do end up making unexpected version changes while this code is still around, I want to be alerted to that fact; as I alluded to in the comment, this logic intentionally maintains a narrow focus.